File: APCS201710_樹的高度與根-2(queue).cpp
#include <bits/stdc++.h>

using namespace std;

#define N 100005
typedef long long ll;

int parent[N], height[N], child[N];

int main()
{
    freopen("P_3_1_5.in", "r", stdin);
    int n, t;
    cin >> n;
    queue<int> q;
    for (int i = 1; i <= n; ++i)
    {
        cin >> child[i];
        if (child[i] == 0)
            q.push(i);
        for (int j = 0; j < child[i]; ++j)
        {
            cin >> t;
            parent[t] = i;
        }
    }
    int root;

    while (!q.empty())
    {
        int now = q.front();
        q.pop();
        int p = parent[now];
        if (p == 0)
        {
            root = now;
            break;
        }
        height[p] = max(height[p], height[now] + 1);
        child[p]--;
        if (child[p] == 0)
            q.push(p);
    }
    cout << root << "\n";
    int sum = 0;
    for (int i = 1; i <= n; ++i)
        sum += height[i];
    cout << sum;
    return 0;
}

File: APCS201710_樹的高度與根-2(queue).py
import sys, queue

# sys.stdin = open("C:\\Users\\ChungHsingHua\\Documents\\c-program\\kevin\\p_7_2_4.in")

n=int(input())

deg=[0]*(n+1)
p=[0]*(n+1)
h=[0]*(n+1)

que=queue.Queue()
for i in range(1,n+1):
    deg[i],*chs=[int(x) for x in input().split()]
    for ch in chs:
        p[ch]=i
    if deg[i]==0:
        que.put(i)

root = p[1:].index(0)+1
print(root)

while not que.empty():
    v = que.get()
    if v == root:
        break
    parent = p[v]
    h[parent]=max(h[v]+1,h[parent])
    deg[parent]-=1
    if deg[parent]==0:
        que.put(parent)
print(sum(h))
